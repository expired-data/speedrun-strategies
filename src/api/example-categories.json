{
  "data": [
    {
      "id": "7dg8wg24",
      "name": "Career 100%",
      "weblink": "https://www.speedrun.com/skate1#Career_100%25",
      "type": "per-game",
      "rules": "End is after you beat Mega-Mayhem.",
      "players": {
        "type": "exactly",
        "value": 1
      },
      "miscellaneous": false,
      "links": [
        {
          "rel": "self",
          "uri": "https://www.speedrun.com/api/v1/categories/7dg8wg24"
        },
        {
          "rel": "game",
          "uri": "https://www.speedrun.com/api/v1/games/vo6gen12"
        },
        {
          "rel": "variables",
          "uri": "https://www.speedrun.com/api/v1/categories/7dg8wg24/variables"
        },
        {
          "rel": "records",
          "uri": "https://www.speedrun.com/api/v1/categories/7dg8wg24/records"
        },
        {
          "rel": "runs",
          "uri": "https://www.speedrun.com/api/v1/runs?category=7dg8wg24"
        },
        {
          "rel": "leaderboard",
          "uri": "https://www.speedrun.com/api/v1/leaderboards/vo6gen12/category/7dg8wg24"
        }
      ]
    },
    {
      "id": "xd1mowd8",
      "name": "X-Games%",
      "weblink": "https://www.speedrun.com/skate1#X-Games",
      "type": "per-game",
      "rules": "End is after you gold medal both X-Games events.",
      "players": {
        "type": "exactly",
        "value": 1
      },
      "miscellaneous": true,
      "links": [
        {
          "rel": "self",
          "uri": "https://www.speedrun.com/api/v1/categories/xd1mowd8"
        },
        {
          "rel": "game",
          "uri": "https://www.speedrun.com/api/v1/games/vo6gen12"
        },
        {
          "rel": "variables",
          "uri": "https://www.speedrun.com/api/v1/categories/xd1mowd8/variables"
        },
        {
          "rel": "records",
          "uri": "https://www.speedrun.com/api/v1/categories/xd1mowd8/records"
        },
        {
          "rel": "runs",
          "uri": "https://www.speedrun.com/api/v1/runs?category=xd1mowd8"
        },
        {
          "rel": "leaderboard",
          "uri": "https://www.speedrun.com/api/v1/leaderboards/vo6gen12/category/xd1mowd8"
        }
      ]
    }
  ]
}
