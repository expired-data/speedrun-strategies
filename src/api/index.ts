export * from "./src";
export * from "./srs";
