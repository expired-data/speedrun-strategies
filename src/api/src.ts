import { useEffect, useState } from "react";

const apiBase = "https://www.speedrun.com/api/v1";

export interface Names {
  international: string;
  japanese: string | null;
  twitch?: string;
}

export interface BulkGame {
  abbreviation: string;
  id: string;
  names: Names;
  weblink: string;
  links: Array<Link<GameLinkType>>;
}

export enum GameLinkType {
  Self = "self",
  Runs = "runs",
  Levels = "levels",
  Categories = "categories",
  Variables = "variables",
  Records = "records",
  Series = "series",
  BaseGame = "base-game",
  DerivedGames = "derived-games",
  RomHacks = "romhacks",
  Leaderboard = "leaderboard",
}

enum PageLinkType {
  Next = "next",
  Previous = "prev",
}

interface Link<LinkType extends string> {
  rel: LinkType;
  uri: string;
}

interface Pagination {
  offset: number;
  max: number;
  size: number;
  links: Array<Link<PageLinkType>>;
}

interface GamesResponseData {
  data: Array<BulkGame>;
  pagination: Pagination;
}

export interface Category {
  id: string;
  name: string;
  weblink: string;
}

interface CategoriesResponseData {
  data: Array<Category>;
}

export interface Time {
  primary: string;
  primary_t: number;
}

export interface UserStub {
  id: string;
  rel: "user";
  uri: string;
}

export interface GuestStub {
  name: string;
  rel: "guest";
  uri: string;
}

export type PlayerStub = UserStub | GuestStub;

export interface Guest {
  name: string;
  rel: "guest";
  uri: string;
}

export interface User {
  id: string;
  names: Names;
  weblink: string;
}

export type Player = Guest | User;

export interface RunType {
  id: string;
  weblink: string;
  game: string;
  level: string | null;
  category: string;
  date: string;
  players: Array<PlayerStub>;
  videos: {
    links: Array<{ uri: string }>;
  };
  comment: string | null;
  times: Time;
}

export interface Run {
  place: number;
  run: RunType;
}

export interface Leaderboard {
  runs: Array<Run>;
  players: { data: Player[] };
}

export const getGames = async (search?: string): Promise<Array<BulkGame>> => {
  let url = `${apiBase}/games?`;

  if (search) {
    url += `&name=${search}`;
  }

  const response: GamesResponseData = await (await fetch(url)).json();

  return response.data;
};

export const getCategories = async (id: string): Promise<Array<Category>> => {
  const url = `${apiBase}/games/${id}/categories`;

  const response: CategoriesResponseData = await (await fetch(url)).json();

  return response.data;
};

export const getLeaderboards = async (
  categoryId: string,
  gameId: string
): Promise<Leaderboard> => {
  const url = `${apiBase}/leaderboards/${gameId}/category/${categoryId}?embed=players`;

  const response: { data: Leaderboard } = await (await fetch(url)).json();

  return response.data;
};

/*eslint-disable-next-line @typescript-eslint/no-explicit-any*/
export const useGetData = <T, Args extends any[]>(
  getDataFn: (...args: Args) => Promise<T>,
  ...args: Args
): T | undefined => {
  const [data, setData] = useState<T | undefined>();

  useEffect(() => {
    setData(undefined);
    getDataFn(...args).then(setData);
  }, [...args]);

  return data;
};
