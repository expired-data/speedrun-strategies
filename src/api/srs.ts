import { gql, useMutation, useQuery } from "@apollo/client";
import { useEffect, useMemo, useState } from "react";
import type { Player, RunType } from "./src";

export interface Strat {
  id: string;
  run: string;
  timestamp: number;
  name: string;
  comment: string;
}

export type PlayerRunType = Omit<RunType, "players"> & {
  players: Player[];
  place: number;
  strats: Strat[];
};

export interface StratAPI {
  strats: Strat[];
  addStrat: (strat: Strat) => void;
  deleteStrat: (strat: Strat) => void;
}

const getStrats = (run: string) => gql`
  query getStrats {
    strats(where: { run: { _eq: "${run}" } }) {
      comment
      id
      name
      run
      timestamp
    }
  }
`;

const addStratMutation = gql`
  mutation addStrat(
    $comment: String = ""
    $name: String = ""
    $run: String = ""
    $timestamp: Int = 0
  ) {
    insert_strats(
      objects: {
        comment: $comment
        name: $name
        run: $run
        timestamp: $timestamp
      }
    ) {
      returning {
        id
      }
    }
  }
`;

const deleteStratMutation = gql`
  mutation deleteStrat($id: uuid = "") {
    delete_strats(where: { id: { _eq: $id } }) {
      affected_rows
    }
  }
`;

export const useGetStrats = (runId: string): StratAPI => {
  const [strats, setStrats] = useState<Strat[]>([]);
  const stratsQuery = useMemo(() => getStrats(runId), [runId]);
  const { data } = useQuery(stratsQuery);
  useEffect(() => {
    if (data?.strats) {
      setStrats(data.strats);
    }
  }, [data]);
  const [addStratMut] = useMutation(addStratMutation);
  const [deleteStratMut] = useMutation(deleteStratMutation);
  const addStrat = (strat: Strat) => {
    addStratMut({
      variables: {
        comment: strat.comment,
        name: strat.name,
        run: strat.run,
        timestamp: strat.timestamp,
      },
    }).then(({ data }) => {
      if (data.insert_strats?.returning[0]?.id) {
        setStrats((strats) => [
          ...strats,
          { ...strat, id: data.insert_strats?.returning[0]?.id },
        ]);
      }
    });
  };

  const deleteStrat = (strat: Strat) => {
    deleteStratMut({
      variables: {
        id: strat.id,
      },
    });
    setStrats((strats) => strats.filter((s) => s.id !== strat.id));
  };

  return {
    strats: strats,
    addStrat,
    deleteStrat,
  };
};
