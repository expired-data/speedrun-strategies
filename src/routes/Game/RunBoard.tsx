import React, { FC, useMemo } from "react";
import { useHistory } from "react-router-dom";

import {
  PlayerRunType,
  useGetData,
  GuestStub,
  UserStub,
  Guest,
  User,
  getLeaderboards,
  Player,
} from "api";
import { splitTime } from "utils/time";
import { Column } from "react-table";
import { Table } from "components/Table";

export interface Props {
  categoryId: string;
  gameId: string;
  onlyWithStrats?: boolean;
}

export const RunBoard: FC<Props> = ({ categoryId, gameId, onlyWithStrats }) => {
  const history = useHistory();
  const leaderboard = useGetData(getLeaderboards, categoryId, gameId);

  const runsWithPlayerNames: PlayerRunType[] = useMemo(
    () =>
      leaderboard?.runs
        .map((run) => ({
          ...run.run,
          players: run.run.players.map((player) => {
            let resultantPlayer: Player | undefined;
            if ((player as UserStub).id) {
              resultantPlayer = leaderboard.players.data.find(
                (fullPlayer) =>
                  (fullPlayer as User).id === (player as UserStub).id
              );
            } else {
              resultantPlayer = leaderboard.players.data.find(
                (fullPlayer) =>
                  (fullPlayer as Guest).name === (player as GuestStub).name
              );
            }

            if (!resultantPlayer) {
              throw new Error("Unable to find player name");
            }

            return resultantPlayer;
          }),
          place: run.place,
          strats: [],
        }))
        .filter((run) => !onlyWithStrats || run.strats.length > 0) || [],
    [leaderboard, onlyWithStrats]
  );

  const columns: Column<PlayerRunType>[] = useMemo(
    () => [
      {
        accessor: "place",
        Header: "Position",
      },
      {
        accessor: "times",
        Cell: ({ value: { primary: runTime } }) => <>{splitTime(runTime)}</>,
        Header: "Time",
      },
      {
        accessor: "players",
        Cell: ({ value: players }) => (
          <>
            {players
              .map(
                (player) =>
                  (player as Guest).name || (player as User).names.international
              )
              .join(", ")}
          </>
        ),
        Header: "Player",
      },
      {
        accessor: "strats",
        Cell: ({ value: strats }) => <>{strats.length}</>,
        Header: "Number of strats",
      },
    ],
    []
  );

  if (!leaderboard) {
    return <div>Loading...</div>;
  }

  return (
    <Table<PlayerRunType>
      data={runsWithPlayerNames}
      columns={columns}
      rowClick={(row) => {
        history.push(`/run/${row.original.id}`);
      }}
    />
  );
};
