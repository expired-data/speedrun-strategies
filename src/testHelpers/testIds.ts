export const SKATE_3_ID = "9j1l7vdg";
export const SKATE_1_ID = "vo6gen12";
export const ANY_PERCENT_ID = "r02q8y2y";
