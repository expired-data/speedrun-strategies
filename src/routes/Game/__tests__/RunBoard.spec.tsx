import React from "react";
import { render } from "@testing-library/react";

import { getLeaderboards, Leaderboard } from "api";
import { SKATE_3_ID, ANY_PERCENT_ID } from "testHelpers/testIds";
import { MOCK_LEADERBOARD, ERROR_LEADERBOARD } from "./mockData";

import { RunBoard } from "../RunBoard";

console.error = jest.fn();

jest.mock("api", () => ({
  __esModule: true,
  getLeaderboards: jest.fn(
    (categoryId, gameId): Promise<Leaderboard> => {
      return gameId === SKATE_3_ID
        ? Promise.resolve(MOCK_LEADERBOARD)
        : Promise.resolve(ERROR_LEADERBOARD);
    }
  ),
  useGetData: jest.requireActual("api").useGetData,
}));

test("Run board renders with leaderboard from SRC", async () => {
  const { findByRole } = render(
    <RunBoard categoryId={ANY_PERCENT_ID} gameId={SKATE_3_ID} />
  );
  const leaderboard = await findByRole("table");
  expect(getLeaderboards).toHaveBeenCalled();
  expect(leaderboard.innerHTML).toMatchSnapshot();
});

test("Run board renders with non strat runs", async () => {
  const { findByRole } = render(
    <RunBoard
      categoryId={ANY_PERCENT_ID}
      gameId={SKATE_3_ID}
      onlyWithStrats={false}
    />
  );
  const leaderboard = await findByRole("table");
  expect(getLeaderboards).toHaveBeenCalled();
  expect(leaderboard.innerHTML).toMatchSnapshot();
});

test("Run board errors if cant find player SRC", async () => {
  expect(async () => {
    const { findByRole } = render(
      <RunBoard categoryId={ANY_PERCENT_ID} gameId={""} />
    );
    await findByRole("table");
    expect(console.error).toHaveBeenCalledTimes(1);
  }).rejects.toThrowError();
});
