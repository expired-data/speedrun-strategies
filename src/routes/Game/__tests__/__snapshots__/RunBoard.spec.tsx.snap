// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Run board renders with leaderboard from SRC 1`] = `"<thead><tr role=\\"row\\" class=\\"sc-gsTCUz kGUbwO\\"><th colspan=\\"1\\" role=\\"columnheader\\">Position</th><th colspan=\\"1\\" role=\\"columnheader\\">Time</th><th colspan=\\"1\\" role=\\"columnheader\\">Player</th><th colspan=\\"1\\" role=\\"columnheader\\">Number of strats</th></tr></thead><tbody role=\\"rowgroup\\"><tr role=\\"row\\" class=\\"sc-gsTCUz dGdWN\\"><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">1</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">1h 07m 46s</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">expireddata</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">1</td></tr><tr role=\\"row\\" class=\\"sc-gsTCUz eHhEwV\\"><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">2</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">1h 13m 52s</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">Mattie</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">0</td></tr><tr role=\\"row\\" class=\\"sc-gsTCUz dGdWN\\"><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">3</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">1h 14m 53s</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">Zockende_Socke</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">0</td></tr></tbody>"`;

exports[`Run board renders with non strat runs 1`] = `"<thead><tr role=\\"row\\" class=\\"sc-gsTCUz kGUbwO\\"><th colspan=\\"1\\" role=\\"columnheader\\">Position</th><th colspan=\\"1\\" role=\\"columnheader\\">Time</th><th colspan=\\"1\\" role=\\"columnheader\\">Player</th><th colspan=\\"1\\" role=\\"columnheader\\">Number of strats</th></tr></thead><tbody role=\\"rowgroup\\"><tr role=\\"row\\" class=\\"sc-gsTCUz dGdWN\\"><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">1</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">1h 07m 46s</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">expireddata</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">1</td></tr><tr role=\\"row\\" class=\\"sc-gsTCUz eHhEwV\\"><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">2</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">1h 13m 52s</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">Mattie</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">0</td></tr><tr role=\\"row\\" class=\\"sc-gsTCUz dGdWN\\"><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">3</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">1h 14m 53s</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">Zockende_Socke</td><td role=\\"cell\\" class=\\"sc-dlfnbm hWxGDI\\">0</td></tr></tbody>"`;
